test: Use MiniWallet in mempool_accept.py

sriramdvt opened the pull request:
Replace all wallet-related functionality in `test/functional/mempool_accept.py` to use MiniWallet instead of the wallet built with Bitcoin Core. This allows the test to run even if Bitcoin Core was compiled with `--disable-wallet`.

Work on `mempool_accept.py` started in #21014, but it has been inactive for some time. This PR also makes use of additional features like `scan_blocks()` and `create_self_transfer()` that were added to MiniWallet.

To test this PR, build Bitcoin Core with(out) the wallet and run:
```
$ test/functional/mempool_accept.py
```

theStack commented:
Concept ACK

theStack commented:
@sriramdvt: Your PR was fine as it is -- #22378 is not merged into master yet, so there is no need to resolve a conflict by now; you can remove the second commit. The DrahtBot conflict message above just says that whenever either your PR or #22378 is merged, the other one very likely has to be rebased.

MarcoFalke commented:
Concept ACK.

In the commit message, the `Co-authored-by:` needs to be a trailer (last line of the message).
( https://docs.github.com/en/github/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors )

theStack commented on test/functional/mempool_accept.py:
```diff
@@ -72,85 +69,71 @@ def run_test(self):
+        coin = miniwallet.get_utxo()  # Get a utxo and store it for later
+        raw_tx_0 = miniwallet.create_self_transfer(fee_rate=fee, from_node=node, utxo_to_spend=coin, sequence=BIP125_SEQUENCE_NUMBER) # RBF is used later
+        tx = tx_from_hex(raw_tx_0['hex'])
         txid_0 = tx.rehash()
```
nit: I think the txid can simpy be obtained by accessing the dict field `txid` of the self-transfer method's result (rehashing shouldn't be needed here).

sriramdvt replied:
Removed multiple instances where rehashing the transaction was not required. `tx.rehash()` is now used only when a `CTransaction()` object is changed.

theStack commented on test/functional/mempool_accept.py:
```diff
@@ -72,85 +69,71 @@ def run_test(self):
-        ))['hex']
-        tx = tx_from_hex(raw_tx_final)
-        fee_expected = coin['amount'] - output_amount
+        raw_tx_final = miniwallet.create_self_transfer(from_node=node, sequence=0xffffffff, locktime=node.getblockcount()+2000) # SEQUENCE_FINAL, locktime can be anything
```
stylistic nit, to better match comments to parameters:

sriramdvt replied:
This was indeed how the first commit (aa068ad397cfe5766dccfd45c0f4ba846e7abd22) had it, but I think I missed it in later force pushes. Thank you!

theStack reviewed (commented):
Left some review comments below, most concerning variable naming and stylistic improvements. Probably the changes in MiniWallet (comment fix in `scan_blocks`, returning fee in `create_self_transfer`) deserve an own commit ahead of the changes in `mempool_accept.py`.

sriramdvt committed:
Minor changes to MiniWallet

sriramdvt committed:
Run mempool_accept.py with MiniWallet

Replaced all wallet functionality in `test/functional/mempool_accept.py` to use `test/functional/test_framework/wallet.py` instead of the wallet built with Bitcoin Core.

Co-authored-by: Sishir Giri <sishirg27@gmail.com>
Co-authored-by: Sebastian Falbesoner <sebastian.falbesoner@gmail.com>

sriramdvt commented on test/functional/mempool_accept.py:
```diff
@@ -72,85 +69,71 @@ def run_test(self):
-            outputs=[{node.getnewaddress(): 0.3}, {node.getnewaddress(): 49}],
-        ))['hex']
-        txid_in_block = node.sendrawtransaction(hexstring=raw_tx_in_block, maxfeerate=0)
+        raw_tx_in_block = miniwallet.send_self_transfer(from_node=node)
```
Removed all instances where having a `raw_` prefix is misleading. There is one instance where `raw_tx_coinbase_spent` gets a raw transaction from `node.getrawtransaction()`. This is the only transaction variable that has a `raw_` prefix now.

sriramdvt commented on test/functional/mempool_accept.py:
```diff
@@ -72,85 +69,71 @@ def run_test(self):
-            outputs=[{node.getnewaddress(): Decimal('0.3') - fee}],
-        ))['hex']
-        tx = tx_from_hex(raw_tx_0)
+        fee = Decimal("0.003")  # The default fee_rate in MiniWallet
```
Changed `fee` to `fee_rate`, thank you!

sriramdvt commented on test/functional/mempool_accept.py:
```diff
@@ -72,85 +69,71 @@ def run_test(self):
+        fee = Decimal("0.003")  # The default fee_rate in MiniWallet
+        coin = miniwallet.get_utxo()  # Get a utxo and store it for later
+        raw_tx_0 = miniwallet.create_self_transfer(fee_rate=fee, from_node=node, utxo_to_spend=coin, sequence=BIP125_SEQUENCE_NUMBER) # RBF is used later
+        tx = tx_from_hex(raw_tx_0['hex'])
```
`tx_from_hex` is now used only when the previous transaction from `{create,send}_self_transfer()` has been modified, or when the test needs to construct a transaction from the raw transaction from `node.getrawtransaction()`.

practicalswift commented:
Concept ACK

More MiniWallet is better.

MarcoFalke commented on test/functional/mempool_accept.py:
```diff
@@ -159,64 +139,53 @@ def run_test(self):
+        node.sendrawtransaction(hexstring=tx_0['hex'], maxfeerate=0) # spend the tx_0
+        self.mempool_size += 1
+        tx_1 = miniwallet.send_self_transfer(fee_rate=4*fee_rate,from_node=node)
+        utxo_1 = miniwallet.get_utxo()
+        miniwallet.send_self_transfer(from_node=node, utxo_to_spend=utxo_1)['txid']
```
```suggestion
        miniwallet.send_self_transfer(from_node=node)['txid']
```

should be identical

MarcoFalke replied:
Also, why the `['txid']` in the end?

sriramdvt replied:
Removed the unnecessary `['txid']`. The suggestion is not identical, since `miniwallet.send_self_transfer` sorts the utxo set by the value, before choosing a transaction.

MarcoFalke commented on test/functional/mempool_accept.py:
```diff
@@ -159,64 +139,53 @@ def run_test(self):
-        ))['hex']
-        tx = tx_from_hex(raw_tx_reference)
+        utxo_2 = miniwallet.get_utxo()
+        tx_reference = miniwallet.create_self_transfer(from_node=node, utxo_to_spend=utxo_2)
```
same

MarcoFalke commented on test/functional/mempool_accept.py:
```diff
@@ -159,64 +139,53 @@ def run_test(self):
-        tx = tx_from_hex(raw_tx_reference)
+        tx = tx_from_hex(tx_reference['hex'])
         tx.vout = []
         # Skip re-signing the transaction for context independent checks from now on
```
Can also remove this line, because there are no signatures

sriramdvt replied:
`tx.vout = []` is necessary since the transaction is serialized in `check_mempool_result`. Without this, the transaction does not get rejected for the right reasons. Please let me know if I misunderstood the suggestion.

MarcoFalke replied:
I meant the comment line :sweat_smile:

MarcoFalke commented on test/functional/mempool_accept.py:
```diff
@@ -159,64 +139,53 @@ def run_test(self):
 
         self.log.info('A transaction with no outputs')
-        tx = tx_from_hex(raw_tx_reference)
+        tx = tx_from_hex(tx_reference['hex'])
```
wouldn't the diff be smaller if the `raw_tx_reference` symbol was kept?

sriramdvt replied:
Agreed! Added a `raw_tx_reference` symbol and the diff is much smaller now.

maflcko reviewed (commented):
Left some comments

sriramdvt commented:
Rebased and updated with the suggestions

laanwj commented:
Unfortunately this needs significant changes due to changes to the affected tests since. Are you still planning to address this?

sriramdvt commented:
@laanwj Thank you for the reminder! I've edited it to be a draft PR in the meanwhile, I will address the changes in a few days and change the status.

theStack commented:
mempool_accept.py is already using MiniWallet on master, i.e. this PR should be closed.
